### PR TITLE
[azure-arm] add controlled storage account name generation

### PR DESCRIPTION
**Pull request by andreaturli**

### CODE REVIEW

File: azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/util/DeploymentTemplateBuilder.java [hunk 1]
```diff
@@ -487,4 +496,37 @@ private boolean keyVaultInUse(){
...
+    *
+    * Storage account names must be between 3 and 24 characters in length and may contain numbers and lowercase letters only.
+    *
+    * @param name the node name
+    * @return the storage account name starting from a sanitized name (with only numbers and lowercase letters only ).
+    * If sanitized name is between 3 and 24 characters, storage account name is equals to sanitized name.
+    * If sanitized name is less than 3 characters, storage account is sanitized name plus 4 random chars.
+    * If sanitized name is more than 24 characters, storage account is first 10 chars of sanitized name plus 4 random chars plus last 10 chars of sanitized name.
+    */
+   private static String generateStorageAccountName(String name) {
```
Comment (nacx) [inline azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/util/DeploymentTemplateBuilder.java]: Please, add appropriate unit tests for this method so we don't break it unintentionally in the future. Tests should properly test the naming convention and the boundaries.

File: azurecompute-arm/src/test/java/org/jclouds/azurecompute/arm/features/DeploymentTemplateBuilderTest.java [hunk 2]
```diff
@@ -132,6 +132,41 @@ void testVirtualMachine() {
...
+   void testAddStorageResourceWhenNameIsLongerThan24Chars() {
+      String name = "thishasmorethan24characters";
+      DeploymentTemplateBuilder builder = getMockDeploymentTemplateBuilderWithEmptyOptions(name);
+
+      DeploymentBody deploymentBody = builder.getDeploymentTemplate();
+      assertTrue(Iterables.contains(deploymentBody.template().variables().keySet(), "storageAccountName"));
+      String storageAccountName = deploymentBody.template().variables().get("storageAccountName");
+      assertEquals(storageAccountName.length(), 24);
+      assertEquals(storageAccountName.substring(0, 10), name.replaceAll("[^a-z0-9]", "").substring(0, 10));
+      assertEquals(storageAccountName.substring(storageAccountName.length() - 10, storageAccountName.length()), name.replaceAll("[^a-z0-9]", "").substring(name.length() - 10, name.length()));
```
Comment (nacx) [inline azurecompute-arm/src/test/java/org/jclouds/azurecompute/arm/features/DeploymentTemplateBuilderTest.java]: Tests should not replicate the business logic (or they aren't actually testing that it produces the expected value; if the business logic is broken, the test is broken too).
Instead, make assertions against concrete values. Assert that the value is a concrete one, or starts with the expected String and has N prefix and N suffix characters, etc.

**Comment by andreaturli** [comment]
@ritazh can you have a look at this PR?

**Comment by ritazh** [comment]
@andreaturli this is awesome! I will test it soon.

**Comment by ritazh** [comment]
@andreaturli LGTM!

**Comment by andreaturli** [comment]
thx @ritazh merging it now

**Comment by andreaturli** [comment]
Merged at [master](http://git-wip-us.apache.org/repos/asf/jclouds-labs/commit/d0b07a66)

**Pull request closed without merging**